src/test/java/com/nimbusds/jose/crypto/ECDH1PUCryptoTest.java

```diff
@@ -0,0 +1,361 @@
+			ECKey bobKey = generateECJWK(cycle.curve);
+
+			JWEHeader header = new JWEHeader.Builder(cycle.algorithm, cycle.encryptionMethod).
+					agreementPartyUInfo(Base64URL.encode("Alice")).
+					agreementPartyVInfo(Base64URL.encode("Bob")).
+					build();
+
+			JWEObject jweObject = new JWEObject(header, payload);
+
+			ECDH1PUEncrypter encrypter = new ECDH1PUEncrypter(aliceKey.toECPrivateKey(), bobKey.toECPublicKey(), cek);
+			encrypter.getJCAContext().setContentEncryptionProvider(BouncyCastleProviderSingleton.getInstance());
+			jweObject.encrypt(encrypter);
+
+			ECKey epk = (ECKey) jweObject.getHeader().getEphemeralPublicKey();
+			assertEquals(cycle.curve, epk.getCurve());
+			assertNotNull(epk.getX());
+			assertNotNull(epk.getY());
+			assertNull(epk.getD());
+
+			String jwe = jweObject.serialize();
+			jweObject = JWEObject.parse(jwe);
+
+			ECDH1PUDecrypter decrypter = new ECDH1PUDecrypter(bobKey.toECPrivateKey(), aliceKey.toECPublicKey());
+			decrypter.getJCAContext().setContentEncryptionProvider(BouncyCastleProviderSingleton.getInstance());
+			jweObject.decrypt(decrypter);
+
+			assertEquals(payload.toString(), jweObject.getPayload().toString());
+		}
+	}
+
+	public void testCritParamDeferral()
+		throws Exception {
+
+		ECKey aliceKey = generateECJWK(Curve.P_256);
+		ECKey bobKey = generateECJWK(Curve.P_256);
+
+		JWEHeader header = new JWEHeader.Builder(JWEAlgorithm.ECDH_1PU, EncryptionMethod.A128CBC_HS256).
+			customParam("exp", "2014-04-24").
+			criticalParams(new HashSet<>(Collections.singletonList("exp"))).
+			build();
+
+		JWEObject jweObject = new JWEObject(header, new Payload("Hello world!"));
+		jweObject.encrypt(new ECDH1PUEncrypter(aliceKey.toECPrivateKey(), bobKey.toECPublicKey()));
+
+		jweObject = JWEObject.parse(jweObject.serialize());
+
+		jweObject.decrypt(new ECDH1PUDecrypter(
+				bobKey.toECPrivateKey(),
+				aliceKey.toECPublicKey(),
+				new HashSet<>(Collections.singletonList("exp"))));
+
+		assertEquals("Hello world!", jweObject.getPayload().toString());
+	}
+
+
+	public void testCritParamReject()
+		throws Exception {
+
+		ECKey aliceKey = generateECJWK(Curve.P_256);
+		ECKey bobKey = generateECJWK(Curve.P_256);
+
+		JWEHeader header = new JWEHeader.Builder(JWEAlgorithm.ECDH_1PU, EncryptionMethod.A128CBC_HS256).
+			customParam("exp", "2014-04-24").
+			criticalParams(new HashSet<>(Collections.singletonList("exp"))).
+			build();
+
+		JWEObject jweObject = new JWEObject(header, new Payload("Hello world!"));
+		jweObject.encrypt(new ECDH1PUEncrypter(aliceKey.toECPrivateKey(), bobKey.toECPublicKey()));
+
+		jweObject = JWEObject.parse(jweObject.serialize());
+
+		try {
+			jweObject.decrypt(new ECDH1PUDecrypter(bobKey.toECPrivateKey(), aliceKey.toECPublicKey()));
+			fail();
+		} catch (JOSEException e) {
+			// ok
+			assertEquals("Unsupported critical header parameter(s)", e.getMessage());
+		}
+	}
+
+	public void testCurveNotMatch() throws Exception {
+		ECKey aliceKey = generateECJWK(Curve.P_256);
```

Comment:
Can we have a test where the key types match each other, but the key itself (points on the curve) don't match? Just to make sure that check not just a type string (which will be set by attacker to the expected value), but the points as well?

Reply:
I'm not able to figure out a test case, becouse a key are parsed by JWK, and JWK already has this checks

Reply:
OK, if Nimbus has other tests that check that it's not possible to create such invalid keys (exception is raised) - then fine.